### Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

### Loading Datasets

In [2]:
df1 = pd.read_csv('train_merc.csv')
df2 = pd.read_csv('test_merc.csv')

In [3]:
print(df1.head())
print(df2.head())

   ID       y  X0 X1  X2 X3 X4 X5 X6 X8  ...  X375  X376  X377  X378  X379  \
0   0  130.81   k  v  at  a  d  u  j  o  ...     0     0     1     0     0   
1   6   88.53   k  t  av  e  d  y  l  o  ...     1     0     0     0     0   
2   7   76.26  az  w   n  c  d  x  j  x  ...     0     0     0     0     0   
3   9   80.62  az  t   n  f  d  x  l  e  ...     0     0     0     0     0   
4  13   78.02  az  v   n  f  d  h  d  n  ...     0     0     0     0     0   

   X380  X382  X383  X384  X385  
0     0     0     0     0     0  
1     0     0     0     0     0  
2     0     1     0     0     0  
3     0     0     0     0     0  
4     0     0     0     0     0  

[5 rows x 378 columns]
   ID  X0 X1  X2 X3 X4 X5 X6 X8  X10  ...  X375  X376  X377  X378  X379  X380  \
0   1  az  v   n  f  d  t  a  w    0  ...     0     0     0     1     0     0   
1   2   t  b  ai  a  d  b  g  y    0  ...     0     0     1     0     0     0   
2   3  az  v  as  f  d  a  j  j    0  ...     0     0     0 

In [4]:
print(df1.shape)
print(df2.shape)

(4209, 378)
(4209, 377)


In [5]:
print(df1.columns)
print(df2.columns)

Index(['ID', 'y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=378)
Index(['ID', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=377)


### 1. If for any column(s), the variance is equal to zero, then you need to remove those variable(s).

In [6]:
df_train = df1.drop('ID',axis=1)
df_test = df2.drop('ID',axis=1)

Since the 'ID' column does not effect the predictions of the dataset, hence were dropped from train and test datasets.

In [7]:
zero_var = df_train.var()[df_train.var()==0].index.values

In [8]:
zero_var

array(['X11', 'X93', 'X107', 'X233', 'X235', 'X268', 'X289', 'X290',
       'X293', 'X297', 'X330', 'X347'], dtype=object)

In [9]:
df_train1 = df_train.drop(zero_var,axis = 1)
df_test1 = df_test.drop(zero_var,axis = 1)

In [10]:
print(df_train1.columns)
print(df_test1.columns)

Index(['y', 'X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=365)
Index(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8', 'X10', 'X12',
       ...
       'X375', 'X376', 'X377', 'X378', 'X379', 'X380', 'X382', 'X383', 'X384',
       'X385'],
      dtype='object', length=364)


### 2. Check for null and unique values for test and train sets.

In [11]:
print(np.sum(df_train1.isna().sum()))
print(np.sum(df_test1.isna().sum()))

0
0


In [12]:
for j in df_train1.columns:
    print(df_train1[j].unique())

[130.81  88.53  76.26 ...  85.71 108.77  87.48]
['k' 'az' 't' 'al' 'o' 'w' 'j' 'h' 's' 'n' 'ay' 'f' 'x' 'y' 'aj' 'ak' 'am'
 'z' 'q' 'at' 'ap' 'v' 'af' 'a' 'e' 'ai' 'd' 'aq' 'c' 'aa' 'ba' 'as' 'i'
 'r' 'b' 'ax' 'bc' 'u' 'ad' 'au' 'm' 'l' 'aw' 'ao' 'ac' 'g' 'ab']
['v' 't' 'w' 'b' 'r' 'l' 's' 'aa' 'c' 'a' 'e' 'h' 'z' 'j' 'o' 'u' 'p' 'n'
 'i' 'y' 'd' 'f' 'm' 'k' 'g' 'q' 'ab']
['at' 'av' 'n' 'e' 'as' 'aq' 'r' 'ai' 'ak' 'm' 'a' 'k' 'ae' 's' 'f' 'd'
 'ag' 'ay' 'ac' 'ap' 'g' 'i' 'aw' 'y' 'b' 'ao' 'al' 'h' 'x' 'au' 't' 'an'
 'z' 'ah' 'p' 'am' 'j' 'q' 'af' 'l' 'aa' 'c' 'o' 'ar']
['a' 'e' 'c' 'f' 'd' 'b' 'g']
['d' 'b' 'c' 'a']
['u' 'y' 'x' 'h' 'g' 'f' 'j' 'i' 'd' 'c' 'af' 'ag' 'ab' 'ac' 'ad' 'ae'
 'ah' 'l' 'k' 'n' 'm' 'p' 'q' 's' 'r' 'v' 'w' 'o' 'aa']
['j' 'l' 'd' 'h' 'i' 'a' 'g' 'c' 'k' 'e' 'f' 'b']
['o' 'x' 'e' 'n' 's' 'a' 'h' 'p' 'm' 'k' 'd' 'i' 'v' 'j' 'b' 'q' 'w' 'g'
 'y' 'l' 'f' 'u' 'r' 't' 'c']
[0 1]
[0 1]
[1 0]
[0 1]
[0 1]
[0 1]
[0 1]
[1 0]
[0 1]
[0 1]
[1 0]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]


Conclusion 1: There are 10 unique values in the Train dataset and the rest are zeros and ones.

In [13]:
for k in df_test1.columns:
    print(df_test1[k].unique())

['az' 't' 'w' 'y' 'x' 'f' 'ap' 'o' 'ay' 'al' 'h' 'z' 'aj' 'd' 'v' 'ak'
 'ba' 'n' 'j' 's' 'af' 'ax' 'at' 'aq' 'av' 'm' 'k' 'a' 'e' 'ai' 'i' 'ag'
 'b' 'am' 'aw' 'as' 'r' 'ao' 'u' 'l' 'c' 'ad' 'au' 'bc' 'g' 'an' 'ae' 'p'
 'bb']
['v' 'b' 'l' 's' 'aa' 'r' 'a' 'i' 'p' 'c' 'o' 'm' 'z' 'e' 'h' 'w' 'g' 'k'
 'y' 't' 'u' 'd' 'j' 'q' 'n' 'f' 'ab']
['n' 'ai' 'as' 'ae' 's' 'b' 'e' 'ak' 'm' 'a' 'aq' 'ag' 'r' 'k' 'aj' 'ay'
 'ao' 'an' 'ac' 'af' 'ax' 'h' 'i' 'f' 'ap' 'p' 'au' 't' 'z' 'y' 'aw' 'd'
 'at' 'g' 'am' 'j' 'x' 'ab' 'w' 'q' 'ah' 'ad' 'al' 'av' 'u']
['f' 'a' 'c' 'e' 'd' 'g' 'b']
['d' 'b' 'a' 'c']
['t' 'b' 'a' 'z' 'y' 'x' 'h' 'g' 'f' 'j' 'i' 'd' 'c' 'af' 'ag' 'ab' 'ac'
 'ad' 'ae' 'ah' 'l' 'k' 'n' 'm' 'p' 'q' 's' 'r' 'v' 'w' 'o' 'aa']
['a' 'g' 'j' 'l' 'i' 'd' 'f' 'h' 'c' 'k' 'e' 'b']
['w' 'y' 'j' 'n' 'm' 's' 'a' 'v' 'r' 'o' 't' 'h' 'c' 'k' 'p' 'u' 'd' 'g'
 'b' 'q' 'e' 'l' 'f' 'i' 'x']
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[0 1]
[1 0]
[1 0]
[1 0]
[0 1]
[

Conclusion 2: There are 10 unique values in the Train and Test datasets and the rest are zeros and ones.

### 3. Apply label encoder.

In [14]:
from sklearn.preprocessing import LabelEncoder

In [15]:
label_col = df_train1.describe(include=['object']).columns.values
label_col

array(['X0', 'X1', 'X2', 'X3', 'X4', 'X5', 'X6', 'X8'], dtype=object)

In [16]:
le = LabelEncoder()

for col in label_col:
    le.fit(df_train1[col].append(df_test1[col]).values)
    df_train1[col]= le.transform(df_train1[col])
    df_test1[col]= le.transform(df_test1[col])

In [17]:
print(df_train1.head())
print(df_test1.head())

        y  X0  X1  X2  X3  X4  X5  X6  X8  X10  ...  X375  X376  X377  X378  \
0  130.81  37  23  20   0   3  27   9  14    0  ...     0     0     1     0   
1   88.53  37  21  22   4   3  31  11  14    0  ...     1     0     0     0   
2   76.26  24  24  38   2   3  30   9  23    0  ...     0     0     0     0   
3   80.62  24  21  38   5   3  30  11   4    0  ...     0     0     0     0   
4   78.02  24  23  38   5   3  14   3  13    0  ...     0     0     0     0   

   X379  X380  X382  X383  X384  X385  
0     0     0     0     0     0     0  
1     0     0     0     0     0     0  
2     0     0     1     0     0     0  
3     0     0     0     0     0     0  
4     0     0     0     0     0     0  

[5 rows x 365 columns]
   X0  X1  X2  X3  X4  X5  X6  X8  X10  X12  ...  X375  X376  X377  X378  \
0  24  23  38   5   3  26   0  22    0    0  ...     0     0     0     1   
1  46   3   9   0   3   9   6  24    0    0  ...     0     0     1     0   
2  24  23  19   5   3   0   9   9

Conclusion 3: As shown above, the 'object' type of columns in Test annd Train datasets have been converted into numbers using Label Encoder

### 4. Perform dimensionality reduction.

In [18]:
from sklearn.decomposition import PCA

In [19]:
pca = PCA(n_components=5)

In [20]:
X = df_train1.drop('y',axis=1)
y = df_train1['y']

In [21]:
from sklearn.model_selection import train_test_split

In [22]:
X_train , X_test , y_train , y_test = train_test_split(X,y,test_size=0.2,random_state=10)

In [23]:
pca.fit(X)

PCA(n_components=5)

In [24]:
pca.n_components_

5

In [25]:
pca.explained_variance_ratio_

array([0.40868988, 0.21758508, 0.13120081, 0.10783522, 0.08165248])

In [26]:
np.sum(pca.explained_variance_ratio_)

0.9469634659642336

Conclusion 4: The optimum value of PCA is selected as 5 of the dimensions, which gives 94.6 % of accurate results.

### 5. Predict your test_df values using XGBoost.

In [27]:
pca_X_train = pd.DataFrame(pca.transform(X_train))
pca_X_test = pd.DataFrame(pca.transform(X_test)) 
pca_test = pd.DataFrame(pca.transform(df_test1))

In [28]:
import xgboost as xgb

In [29]:
model = xgb.XGBRegressor(learning_rate=0.1)

In [30]:
model.fit(pca_X_train,y_train)

XGBRegressor(base_score=0.5, booster=None, colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints=None,
             learning_rate=0.1, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
             tree_method=None, validate_parameters=False, verbosity=None)

In [31]:
pred_y_test = model.predict(pca_X_test)

In [32]:
from sklearn.metrics import mean_squared_error

In [33]:
mse = mean_squared_error(y_test,pred_y_test)

In [34]:
print(mse)

94.24855673333212


###### XG Boost prediction using non-pca values

In [35]:
model1 = xgb.XGBRegressor(learning_rate=0.1)

In [36]:
model1.fit(X_train,y_train)

XGBRegressor(base_score=0.5, booster=None, colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints=None,
             learning_rate=0.1, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints=None,
             n_estimators=100, n_jobs=0, num_parallel_tree=1, random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
             tree_method=None, validate_parameters=False, verbosity=None)

In [37]:
pred_y_test1 = model1.predict(X_test)

In [38]:
mse1 = mean_squared_error(y_test,pred_y_test1)

In [39]:
print(mse1)

75.20578928776922


Conclusion 5: PCA of Train data give 94% of accuracy in prediction, while non-pca or original data gives 75% of accuracy in prediction.